aiida_mlip/data/model.py

```diff
@@ -0,0 +1,125 @@
+"""Define Model Data type in AiiDA"""
+
+import hashlib
+from pathlib import Path
+from urllib import request
+from urllib.parse import urlparse
+
+from aiida.orm import SinglefileData
+
+
+def calculate_hash(file):
+    """Function to calculate file hashes"""
+    # Calculate hash
+    sha256 = hashlib.sha256()
+    with open(file, "rb") as f:
+        for data in iter(lambda: f.read(65536), b""):
+            sha256.update(data)
+    file_hash = sha256.hexdigest()
+    return file_hash
+
+
+def check_existing_file(arch_path, file, file_hash, architecture, cache_path):
+    """Function to check if a file already exists and return a file path and architecture"""
+    for existing_file in arch_path.rglob("*"):
+        if existing_file.is_file() and existing_file != file:
+            existing_file_hash = calculate_hash(existing_file)
+            if existing_file_hash == file_hash:
+                ex_file_path = Path(existing_file).parent
+                if arch_path == ex_file_path:
+                    file.unlink()
+                    return existing_file, architecture
+
+                if arch_path == cache_path and arch_path != ex_file_path:
+                    architecture = ex_file_path.name
+                    file.unlink()
+                    return existing_file, architecture
+    return file, architecture
+
+
+class ModelData(SinglefileData):
+    """Class to handle models"""
```

Comment:
It might be worth making it slightly more clear what model refers to (MLIP?), and very roughly what handling involves (loading/saving?).